src/AspNet.Security.OAuth.Shopify/ShopifyAuthenticationHandler.cs

```diff
@@ -153,35 +152,39 @@ protected override string BuildChallengeUrl(AuthenticationProperties properties,
 
             try
             {
-                var shop = Context.Request.Query["shop"];
-                var state = Context.Request.Query["state"];
+                StringValues shopValue = Context.Request.Query["shop"];
+                StringValues stateValue = Context.Request.Query["state"];
+
+                string shop = shopValue.ToString();
 
                 // Shop name must end with myshopify.com
-                if (!shop.ToString().EndsWith(".myshopify.com", StringComparison.InvariantCulture))
+                if (!shop.EndsWith(".myshopify.com", StringComparison.OrdinalIgnoreCase))
                 {
-                    throw new Exception("Unexpected query string.");
+                    throw new InvalidOperationException("shop parameter is malformed. It should end with .myshopify.com");
                 }
 
                 // Strip out the "myshopify.com" suffix
-                shopDns = shop.ToString().Split('.')[0];
+                shopDns = shop.Split('.')[0];
 
                 // Verify that the shop name encoded in "state" matches the shop name we used to
                 // request the token. This probably isn't necessary, but it's an easy extra verification.
-                var authenticationProperties = Options.StateDataFormat.Unprotect(state);
-                if (!authenticationProperties.Items[ShopifyAuthenticationDefaults.ShopNameAuthenticationProperty]
-                    .Equals(shopDns, StringComparison.InvariantCultureIgnoreCase))
+                AuthenticationProperties authenticationProperties = Options.StateDataFormat.Unprotect(stateValue);
+
+                string shopNamePropertyValue = authenticationProperties.Items[ShopifyAuthenticationDefaults.ShopNameAuthenticationProperty];
+
+                if (!shopNamePropertyValue.Equals(shopDns, StringComparison.OrdinalIgnoreCase))
                 {
-                    throw new Exception("Unexpected query string");
+                    throw new InvalidOperationException("shop parameter is not the same who requested the token");
```

Comment:
_"Received shop name does not match the shop name specified in the authentication request."_ or something similar? I think "who" is a bit vague here.